lib-ee/emqx_ee_connector/src/emqx_ee_connector_cassa.erl

```diff
@@ -0,0 +1,412 @@
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_ee_connector_cassa).
+
+-behaviour(emqx_resource).
+
+-include_lib("emqx_connector/include/emqx_connector.hrl").
+-include_lib("emqx_ee_connector/include/emqx_ee_connector.hrl").
+-include_lib("typerefl/include/types.hrl").
+-include_lib("emqx/include/logger.hrl").
+-include_lib("hocon/include/hoconsc.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+%% schema
+-export([roots/0, fields/1]).
+
+%% callbacks of behaviour emqx_resource
+-export([
+    callback_mode/0,
+    on_start/2,
+    on_stop/2,
+    on_query/3,
+    %% TODO: now_supported_now
+    %%on_batch_query/3,
+    on_get_status/2
+]).
+
+%% callbacks of ecpool
+-export([
+    connect/1,
+    prepare_sql_to_conn/2
+]).
+
+%% callbacks for query executing
+-export([query/3, prepared_query/3]).
+
+-export([do_get_status/1]).
+
+-type prepares() :: #{atom() => binary()}.
+-type params_tokens() :: #{atom() => list()}.
+
+-type state() ::
+    #{
+        poolname := atom(),
+        prepare_sql := prepares(),
+        params_tokens := params_tokens(),
+        %% returned by ecql:prepare/2
+        prepare_statement := binary()
+    }.
+
+-define(DEFAULT_SERVER_OPTION, #{default_port => ?CASSANDRA_DEFAULT_PORT}).
+
+%%--------------------------------------------------------------------
+%% schema
+
+roots() ->
+    [{config, #{type => hoconsc:ref(?MODULE, config)}}].
+
+fields(config) ->
+    cassandra_db_fields() ++
+        emqx_connector_schema_lib:ssl_fields() ++
+        emqx_connector_schema_lib:prepare_statement_fields().
+
+cassandra_db_fields() ->
+    [
+        {servers, servers()},
+        {keyspace, fun keyspace/1},
+        {pool_size, fun emqx_connector_schema_lib:pool_size/1},
+        {username, fun emqx_connector_schema_lib:username/1},
+        {password, fun emqx_connector_schema_lib:password/1},
+        {auto_reconnect, fun emqx_connector_schema_lib:auto_reconnect/1}
+    ].
+
+servers() ->
+    Meta = #{desc => ?DESC("servers")},
+    emqx_schema:servers_sc(Meta, ?DEFAULT_SERVER_OPTION).
+
+keyspace(type) -> binary();
+keyspace(desc) -> ?DESC("keyspace");
+keyspace(required) -> true;
+keyspace(_) -> undefined.
+
+%%--------------------------------------------------------------------
+%% callbacks for emqx_resource
+
+callback_mode() -> always_sync.
+
+-spec on_start(binary(), hoconsc:config()) -> {ok, state()} | {error, _}.
+on_start(
+    InstId,
+    #{
+        servers := Servers,
+        keyspace := Keyspace,
+        username := Username,
+        pool_size := PoolSize,
+        ssl := SSL
+    } = Config
+) ->
+    {ok, _} = application:ensure_all_started(ecpool),
+    {ok, _} = application:ensure_all_started(ecql),
```

Comment:
Q: maybe these should be declared as dependencies to the `emqx_ee_connector` app, so they are started before by the application? (I see that `ecql` was already added)

Reply:
sure!